Oracle merge statement test cases for merge update clause

ThanoshanMV opened the pull request:
#9129 

Here's the `MERGE` statement [documentation](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/MERGE.html#GUID-5692CCB7-24D9-4C0E-81A7-A22436DC968F). 

@tristaZero, @wgy8283335  I've added test cases for `merge_update_clause`

Changes proposed in this pull request:
- Added test cases for `mergeUpdateClause`.
- I modified `MERGE` antlr definition for `mergeUpdateClause`: added two new rules `mergeSetAssignmentsClause`, `deleteWhereClause` and removed parenthesis for `hint` and `mergeAssignmentValue`.
- I tried to visit `mergeUpdateClause` in the `OracleDMLStatementSQLVisitor`.
- I chose `SetAssignmentSegment` and `ExpectedSetClause` for the  `mergeSetAssignmentsClause` similar to `setAssignmentsClause` in the `UPDATE` statement. 
- Since `mergeAssignmentValue` contains columns as values, to expect them, I added `ExpectedColumn` in the `ExpectedAssignmentValue.java` and to assert columns, I added `ColumnAssert` in the `AssignmentValueAssert.java`. 

I'll change them according to your feedback.

wgy8283335 commented on shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/statement/dml/MergeStatement.java:
```diff
@@ -37,4 +41,17 @@
     private TableSegment source;
     
     private ExpressionSegment expr;
+    
```
@tristaZero Hi, do you think it is necessary to add the 'setAssignment' and 'where' in MergeStatement?
Related rule as below:
```
mergeSetAssignmentsClause
    : SET mergeAssignment (COMMA_ mergeAssignment)*
    ;

whereClause
    : WHERE expr
    ;
```

tristaZero replied:
Hi,

First, `mergeSetAssignmentsClause` is OK for me, since it is like a definition with a group of `mergeSetAssignment`. Therefore a `merge` main definition will contain some of the keywords and some of the child group definitions, and each group definition is made up of its child definitions. 
On the other hand, you can still put all specific definitions in the main clause, which does not have any group definitions.

Which one is our best choice? I suppose we need to add another factor, i.e., `SQLStatement`. Suppose we want to visit out some specific `statement`, it is necessary to estimate which type of SQL definition of `g4` file is more reasonable and straightforward for visitors.

That's my opinion and hope it can give you some hints for the following issues.

Second, actually, a more fitting member putting into `MergeStatement`, I guess, is the `UpdateStatement` containing `setAssignment` and `where`. The reason is that `OracleUpdateStatement` has existed and owned the entire and complete definition about `update set where`. 
But that will make it more challenging to do unit tests. I suggest you can have a try, and if that frustrates you too much, you can put it aside. Yep, it is optional. :)

@ThanoshanMV 
Hope you can succeed in GSoC 2021 and learn more from this community.  Also @wgy8283335 and I will provide our mentoring as much as we can.

ThanoshanMV commented on shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DMLStatement.g4:
```diff
@@ -198,15 +198,23 @@ usingClause
 
 mergeUpdateClause
-    : WHEN MATCHED THEN UPDATE SET mergeAssignment (COMMA_ mergeAssignment)* whereClause? (DELETE whereClause)?
+    : WHEN MATCHED THEN UPDATE mergeSetAssignmentsClause whereClause? deleteWhereClause?
```
Yes @wgy8283335, `WHEN MATCHED THEN UPDATE SET`  is better. I put `SET` separately in the `mergeSetAssignmentsClause`  because of its rule name only. Since the rule name specifies `SetAssignments` in its name, I just put keyword `SET` inside that rule. 

I'll change that one as: 
```
mergeUpdateClause
    : WHEN MATCHED THEN UPDATE SET mergeSetAssignmentsClause whereClause? deleteWhereClause?
    ;

mergeSetAssignmentsClause 
    : mergeAssignment (COMMA_ mergeAssignment)*
```

ThanoshanMV replied:
Hi @wgy8283335, I've updated `mergeSetAssignmentsClause`.

tristaZero commented:
Hi,

First, `mergeSetAssignmentsClause` is OK for me, since it is like a definition with a group of `mergeSetAssignment`. Therefore a `merge` main definition will contain some of the keywords and some of the child group definitions, and each group definition is made up of its child definitions. 
On the other hand, you can still put all specific definitions in the main clause, which does not have any group definitions.

Which one is our best choice? I suppose we need to add another factor, i.e., `SQLStatement`. Suppose we want to visit out some specific `statement`, it is necessary to estimate which type of SQL definition of `g4` file is more reasonable and straightforward for visitors.

That's my opinion and hope it can give you some hints for the following issues.

Second, actually, a more fitting member putting into `MergeStatement`, I guess, is the `UpdateStatement` containing `setAssignment` and `where`. The reason is that `OracleUpdateStatement` has existed and owned the entire and complete definition about `update set where`. 
But that will make it more challenging to do unit tests. I suggest you can have a try, and if that frustrates you too much, you can put it aside. Yep, it is optional. :)

@ThanoshanMV 
Hope you can succeed in GSoC 2021 and learn more from this community.  Also @wgy8283335 and I will provide our mentoring as much as we can.

ThanoshanMV commented:
Yes @tristaZero, `UpdateStatement` is more fitting member in the `MergeStatement` as it contains both `setAssignment` and `where`. I'll try to do in that way.

ThanoshanMV commented:
Hi @tristaZero, I've added `UpdateStatement` as a member of the `MergeStatement`. I've modified other files to pass test cases. I'll change them according to your feedback.

tristaZero commented:
Hi @ThanoshanMV ,
Looks better. Let us wait for @wgy8283335 's feedback.

wgy8283335 commented on shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/oracle/visitor/statement/impl/OracleDMLStatementSQLVisitor.java:
```diff
@@ -487,6 +492,15 @@ public ASTNode visitMerge(final MergeContext ctx) {
+                result.getUpdate().setWhere((WhereSegment) visit(ctx.mergeUpdateClause().whereClause()));
+            }
+            if (null != ctx.mergeUpdateClause().deleteWhereClause()) {
+                result.getUpdate().setWhere((WhereSegment) visit(ctx.mergeUpdateClause().deleteWhereClause()));
```
@tristaZero @ThanoshanMV  Delete Where is not suitable in the update statement. Maybe should put in a delete statement.

ThanoshanMV replied:
Okay @wgy8283335, I'll put `DeleteStatement` as a member in the `MergeStatement`. So that we can visit `deleteWhereClause` and put it inside `DeleteStatement`'s `WhereSegment`.

ThanoshanMV replied:
@wgy8283335 since we get `WhereSegement` actual values from two statements: `UpdateStatement` and `DeleteStatement`, to expect and assert those `WhereSegement` values, we should have two different `ExpectedWhereClause` right? 

Can we have something like below in the `MergeStatementTestCase`? :

![MergeStatementTestCase](https://user-images.githubusercontent.com/48581379/112149565-247f9c00-8c05-11eb-87cc-8669016a0e5a.png)

wgy8283335 replied:
Yes, I think so.

ThanoshanMV replied:
@wgy8283335 I added `DeleteStatement` as a member in the `MergeStatement` and modified the `MergeStatementTestCase` like above. 

I asserted both `WhereSegements` inside `assertWhereClause` method in `MergeStatementAssert`.

tristaZero commented:
@wgy8283335 @ThanoshanMV Can we merge this one to master branch?

ThanoshanMV commented:
Hi @tristaZero, the new changes need to be reviewed.

Baoanh217 commented on shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DMLStatement.g4:
```diff
@@ -176,7 +176,7 @@ havingClause
 subquery
     : LP_ unionClause RP_
```
.

wgy8283335 replied:
@tristaZero I've finished the review, shall we merge the pull request?
@Baoanh217 leave a '.' in the comment. Shall we ignore the comment, as the below code is not changed in this submit
```
subquery
    : LP_ unionClause RP_
```

tristaZero commented:
Hi @Baoanh217 , thanks for your attention, please have my welcome.
@wgy8283335 Thanks for your approval moving this PR forward.
